### PR TITLE
Towing vehicles

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Towing vehicles"```

#### Purpose of change
Allows you to tow a vehicle with another one, by attaching a tow-line.

this relies on some of the changes in #34616 , and therefore has those commits too.

#### Describe the solution
Add an item that is similar to charging cable in how it attaches to a vehicle. 
When it attaches it sets references to each vehicle to the other, when the puller moves, the pulled vehicle moves too.

#### Describe alternatives you've considered
N/A

#### Testing
Still to do :- 
1)saving/loading - DONE
2)currently there are pointers stored on each vehicle, these need to be managed with care to ensure they are removed and not left dangling when certain things happen - such as destruction of a vehicle, destruction of the tow line part etc. - DONE
3)Make the tow-line have more than 3 "charges" so it can be spooled further out. - DONE
4) tweak the code that makes it follow closely, so that it is smarter. - DONE
5) handle vehicle getting unloaded out of reality bubble etc. - DONE 
6)add weight to the pulling vehicle to represent the added strain of the pulled vehicles weight. - DONE
7) if pulled vehicle has no wheels, make it skid? make it create shallow pits in the ground? - THIS ALREADY HAPPENS AUTOMATICALLY
8) need to handle obstacles interfering with the pulled vehicle, or when too much distance opens up between the vehicles. or when the pulled vehicle is driven away ( need to make it bidirectional somehow maybe? ) - DONE
9) spawn in various places, make it craftable. - DONE

10) handle vehicles splitting / loaded onto racks - DONE

current limitations : 
The order in which you apply the cable matters, first vehicle selected is the puller, second is the pulled, Ima see if I can work around that.
It really is better if you attach the tow-line to the rear of the puller, and the front of the pulled.
Otherwise weird things may happen.

#### Additional context
A gif of it - 
https://gfycat.com/disguisedweepyladybug

### CODE REVIEW

**Comment by chaisawlajatang** [comment]
Oh, wow. WOW

**Comment by Destragon** [comment]
Can you chain them and have another car being pulled by the pulled vehicle or is it limited to just 1?

**Comment by davidpwbrown** [comment]
> Can you chain them and have another car being pulled by the pulled vehicle or is it limited to just 1?

Currently if a vehicle is either a tower, or a towed, then you cant attach another cable to it, I did this deliberately to keep things simple.

But I guess a towee, could also be a tower, and theoretically yes it could chain. Id be wary of it though, it would multiply the chances for things to go wrong.

**Comment by Destragon** [comment]
Will anything funny happen when #33829 gets merged and you tie a car to a helicopter?

**Comment by davidpwbrown** [comment]
> Will anything funny happen when #33829 gets merged and you tie a car to a helicopter?

Quite likely very strange things would happen.

**Comment by davidpwbrown** [comment]
Rebased to latest changes in #34616

**Comment by Destragon** [comment]
Does the towed vehicle keep its momentum when the tower suddenly hits on the breaks or does it also stop immediately?

**Comment by davidpwbrown** [comment]
> Does the towed vehicle keep its momentum when the tower suddenly hits on the breaks or does it also stop immediately?

Theres no momentum, the towed vehicle is kinda driving itself, and when the distance gets less than a certain number, then it stops moving.  So no hard braking and having thw towed vehicle ram into you, not yet anyway :)

File: src/iuse.cpp [hunk 1]
```diff
@@ -8814,6 +8814,125 @@ int iuse::multicooker( player *p, item *it, bool t, const tripoint &pos )
...
+            p->add_msg_if_player( _( "There's no vehicle there." ) );
+            return 0;
+        } else {
+            vehicle *const source_veh = veh_pointer_or_null( vp );
+            if( source_veh && ( source_veh->has_tow_attached() || source_veh->is_towed() ||
+                                source_veh->is_towing() ) ) {
+                p->add_msg_if_player( _( "That vehicle already has a tow-line attached." ) );
+                return 0;
+            }
+            const auto abspos = g->m.getabs( posp );
```
Comment (KorGgenT) [inline src/iuse.cpp]:
```suggestion
            const tripoint &abspos = g->m.getabs( posp );
```

**Comment by RustyBrute** [comment]
Hell yeah! Good job.

File: src/game.cpp [hunk 2]
```diff
@@ -1869,6 +1869,21 @@ static void update_faction_api( npc *guy )
     }
 }
 
+void game::validate_linked_vehicles()
+{
+    for( auto &veh : m.get_vehicles() ) {
```
Comment (KorGgenT) [inline src/game.cpp]: auto!

File: src/vehicle.cpp [hunk 3]
```diff
@@ -757,42 +758,44 @@ void vehicle::drive_to_local_target( const tripoint target, bool follow_protocol
...
-            stop = true;
-            break;
-        }
-        if( elem == g->u.pos().xy() ) {
-            if( follow_protocol || g->u.in_vehicle ) {
-                continue;
-            } else {
-                stop = true;
+    if( !towed ) {
+        for( const auto &pt : points_to_check ) {
```
Comment (KorGgenT) [inline src/vehicle.cpp]: auto!

File: src/vehicle.cpp [hunk 4]
```diff
@@ -757,42 +758,44 @@ void vehicle::drive_to_local_target( const tripoint target, bool follow_protocol
...
-                stop = true;
-                break;
+            bool its_a_pet = false;
+            if( g->critter_at( tripoint( elem, sm_pos.z ) ) ) {
+                npc *guy = g->critter_at<npc>( tripoint( elem, sm_pos.z ) );
+                if( guy && !guy->in_vehicle ) {
+                    stop = true;
+                    break;
+                }
+                for( const auto &p : parts ) {
```
Comment (KorGgenT) [inline src/vehicle.cpp]: auto! also `p` is a terrible variable name. what is it, point? player?

File: src/vehicle.cpp [hunk 5]
```diff
@@ -5558,6 +5603,184 @@ void vehicle::refresh_pivot() const
...
+    }
+    for( const vpart_reference &vp : get_all_parts() ) {
+        const size_t p = vp.part_index();
+        if( vp.part().removed ) {
+            continue;
+        }
+
+        if( part_with_feature( p, "TOW_CABLE", true ) >= 0 ) {
+            if( first_vehicle ) {
+                auto part = &parts[part_with_feature( p, "TOW_CABLE", true )];
```
Comment (KorGgenT) [inline src/vehicle.cpp]: auto!

File: data/json/vehicleparts/vehicle_parts.json [hunk 6]
```diff
@@ -2631,6 +2631,22 @@
...
+    "color": "light_blue",
+    "broken_symbol": "*",
+    "broken_color": "dark_gray",
+    "damage_modifier": 10,
+    "durability": 120,
+    "description": "A tow cable, if the other end was attached to another vehicle, it could pull it.",
+    "item": "tow_cable",
+    "requirements": { "removal": { "time": 500 } },
+    "flags": [ "NOINSTALL", "UNMOUNT_ON_DAMAGE", "TOW_CABLE" ],
+    "breaks_into": [ { "item": "wire", "count": [ 4, 8 ] }, { "item": "plastic_chunk", "count": [ 4, 8 ] } ]
```
Comment (Night-Pryanik) [inline data/json/vehicleparts/vehicle_parts.json]: There is no wire or plastic in crafting recipe of this cable at all. Also I think grip hook could be dropped too.

**Comment by Night-Pryanik** [comment]
Could you please add tow cable to the list of possible alternatives of rope in butchering activity?

**Comment by davidpwbrown** [comment]
> Could you please add tow cable to the list of possible alternatives of rope in butchering activity?

Done

File: data/json/items/vehicle/cables.json [hunk 7]
```diff
@@ -18,6 +18,25 @@
...
+    "type": "TOOL",
+    "id": "tow_cable",
+    "name": "tow cable",
+    "description": "A tow cable, attached to a vehicle, it could be used to pull another vehicle.",
+    "to_hit": -1,
+    "color": "light_blue",
+    "symbol": "&",
+    "material": [ "steel", "plastic" ],
+    "volume": "3000 ml",
+    "weight": "875 g",
```
Comment (mlangsdorf) [inline data/json/items/vehicle/cables.json]: Since we're not doing weight limits yet, this should be 24 kgs of chain and cable with a volume of 3L or so.

File: data/json/recipes/recipe_vehicle.json [hunk 8]
```diff
@@ -56,6 +56,19 @@
...
+    "type": "recipe",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_VEHICLE",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "time": 120000,
+    "autolearn": true,
+    "qualities": [ { "id": "HAMMER", "level": 1 } ],
+    "using": [ [ "welding_standard", 2 ] ],
+    "components": [ [ [ "duct_tape", 50 ] ], [ [ "grip_hook", 1 ] ], [ [ "chain", 2 ] ] ]
```
Comment (mlangsdorf) [inline data/json/recipes/recipe_vehicle.json]: add some "cable" (ie, copper wire for the electrical hookups) and enough chain to get up to 24 kgs of total weight.

time should be 8 hours or so.

File: data/json/items/vehicle/cables.json [hunk 9]
```diff
@@ -18,6 +18,25 @@
     "use_action": "CABLE_ATTACH",
     "flags": [ "CABLE_SPOOL" ]
   },
+  {
+    "type": "TOOL",
+    "id": "tow_cable",
+    "name": "tow cable",
+    "description": "A tow cable, attached to a vehicle, it could be used to pull another vehicle.",
```
Comment (mlangsdorf) [inline data/json/items/vehicle/cables.json]:
```suggestion
    "description": "An extremely heavy duty tow cable made from thick steel wire.  If attached to a vehicle, it could be used to pull another vehicle of any weight.",
```

File: data/json/items/vehicle/cables.json [hunk 10]
```diff
@@ -18,6 +18,25 @@
     "use_action": "CABLE_ATTACH",
     "flags": [ "CABLE_SPOOL" ]
   },
+  {
+    "type": "TOOL",
+    "id": "tow_cable",
```
Comment (mlangsdorf) [inline data/json/items/vehicle/cables.json]:
```suggestion
    "id": "hd_tow_cable",
```

File: data/json/items/vehicle/cables.json [hunk 11]
```diff
@@ -18,6 +18,25 @@
...
+    "name": "tow cable",
```
Comment (mlangsdorf) [inline data/json/items/vehicle/cables.json]:
```suggestion
    "name": "heavy duty tow cable",
```

File: data/json/vehicleparts/vehicle_parts.json (hunk 6 above)
Comment (Night-Pryanik) [reply data/json/vehicleparts/vehicle_parts.json]: I meant add grip hook to the list of what drops when part is broken.

Comment (davidpwbrown) [reply data/json/vehicleparts/vehicle_parts.json]: Ohh lol, sure yes.

File: data/json/recipes/recipe_vehicle.json (hunk 8 above)
Comment (davidpwbrown) [reply data/json/recipes/recipe_vehicle.json]: The tow cable has electrical hookups?

**Comment by paulenka-aleh** [comment]
It looks great to finally see the concept in game. It feels a bit wrong though:

> Towing support
After the collision rework, it should be possible to have two correctly flagged vehicle parts not collide. And it should also be possible to apply force to the trailer to have it follow the towing vehicle.
Added by mlangsdorf

https://github.com/CleverRaven/Cataclysm-DDA/projects/19

**Comment by davidpwbrown** [comment]
> It looks great to finally see the concept in game. It feels a bit wrong though:
> 
> > Towing support
> > After the collision rework, it should be possible to have two correctly flagged vehicle parts not collide. And it should also be possible to apply force to the trailer to have it follow the towing vehicle.
> > Added by mlangsdorf
> 
> https://github.com/CleverRaven/Cataclysm-DDA/projects/19

This is a different route to the same end-goal, what feels wrong exactly?

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/passive-vehicle-movement/22443/4

**Comment by davidpwbrown** [comment]
Been playing with it some more, needs to go back into [WIP] as certain circumstances ( like towing a long bus from the wrong end ) can cause the bus to flip around and ram repeatedly into the back of your car because it thinks its too far away from the towing point.

**Comment by scorpion451** [comment]
Thought on the "revolving bus" thing: the fix here could also fix the long-standing issues with dragging longer carts by points other than the frontmost tile. Maybe something where the vector from the attachment point to the tower becomes the towed vehicle's "forward". (Similar to how bike racks work, the racked vehicle becoming a sort of black box object until it's unmounted)

Also occurs to me that the bus issue makes partial sense from a physics standpoint- You want fixed wheels at the back of a towed object to make that the pivot point in a turn.  You never put steerable wheels on a long trailer, or tow something like a bus backwards for a long distance, because it causes all sorts of problems - fishtailing, jackknifing, multi-lane drifting... A lot of your major semi accidents involve the trailer's wheels losing traction or getting thrown off balance so that they just flail around like a kite-tail. (TL;DR: the bus issue might not be so much a bug as a public service announcement.)

**Comment by davidpwbrown** [comment]
> Thought on the "revolving bus" thing: the fix here could also fix the long-standing issues with dragging longer carts by points other than the frontmost tile. Maybe something where the vector from the attachment point to the tower becomes the towed vehicle's "forward", ignoring the normal steering properties. (Similar to how bike racks work, the racked vehicle becoming a sort of black box object to the physics until it's unmounted)
> 
> Also occurs to me that the bus issue makes partial sense from a physics standpoint- You want fixed wheels at the back of a towed object to make that the pivot point in a turn. You never put steerable wheels on a long trailer, or tow something like a bus backwards for a long distance, because it causes all sorts of problems - fishtailing, jackknifing, multi-lane drifting... A lot of your major semi accidents involve the trailer's wheels losing traction or getting thrown off balance so that the trailer just flails around like a kite-tail. The flipping and crashing into the towing vehicle could just be havok-physics-style rotation flailing exaggerating a physics problem. (I know 180*+ power slides are possible in the game because I've pulled them off before as a Hail Mary maneuver)
> 
> TL;DR: the bus issue might not be so much a bug as a public service announcement.

Yep I had same conclusion, the towing point becomes the vector by which it moves, should be doable.

**Comment by davidpwbrown** [comment]
Experimental update : now when towing line is attached, it works out whether it is pulling the vehicle from the front, the back, or the side. 
This changes how the towing behaviour works, if its from the side, it just displaces it laterally, not ideal.
If its from the front, it "autodrives" and steers towards the towing vehicle.
if its from the rear, it reverses and autodrives towards the towing vehicle, but steering and acceleration reversed.
Seems to work as I wanted, but obviously more polishing will be required post 0.E freeeze I imagine.

**Comment by pjf** [comment]
Applied this to my local game, and had my deathmobile tow a 4x4.

Worked great on roads, but as soon as I got off-road, my deathmobile continued (slowly), but the vehicle it was towing did not.

The deathmobile isn't built for off-road use, so I suspect the weight of both vehicles combined exceeded its engine capacity. However in that case I'd expect the towing vehicle to stop, not the vehicle being towed.

Otherwise this is great. An excellent change for folks who have multiple vehicles and like to relocate.

**Comment by davidpwbrown** [comment]
Rebased, cleaned up, tested, working.
Fixed bug that pjf reported .

Im not sure why the tower is allowed to proceed at velocity 1, when the towed vehicle stops at velocity 1.

I managed to repro this with a bus pulling a bus, and they went off-road, this would be a situation where probably it would be impossible to get traction, so ive detected this scenario and added a message similar to when a vehicle is too heavy for its engines.

Pulling a golf cart with a military cargo truck off-road didnt have this problem, it managed to stay above 1mph.

**Comment by CEG86** [comment]
How do you detach the tow cable?  I had to deconstruct the towed vechicle which resulted in the tow cable disappearing.

**Comment by ZhilkinSerg** [comment]
Try examining vehicle tile with tow attachment.
